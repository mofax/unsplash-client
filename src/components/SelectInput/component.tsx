import "react-dropdown/style.css";

import * as React from "react";
import * as styles from "./styles.css";

import Dropdown, { ReactDropdownProps } from "react-dropdown";

export function SelectInput(props: ReactDropdownProps) {
    return (
        <Dropdown
            className={styles.input}
            placeholderClassName={styles.placeholder}
            menuClassName={styles.menu}
            {...props}
        />
    );
}
