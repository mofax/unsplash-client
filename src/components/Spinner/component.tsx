import * as React from "react";
import * as styles from "./styles.css";

import { classNames } from "../../tools/classnames";

type SpinnerProps = {
    width?: number;
    height?: number;
    dark?: boolean;
};

export function Spinner(props: SpinnerProps) {
    const inline = {
        width: props.width,
        height: props.height,
    };
    const optionalClasses: string[] = [];
    if (props.dark) {
        optionalClasses.push(styles.dark);
    }
    const classes = classNames(styles.spinner, optionalClasses);

    return <span className={classes} style={inline}></span>;
}

Spinner.defaultProps = {
    width: 16,
    height: 16,
};
