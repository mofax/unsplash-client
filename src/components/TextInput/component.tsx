import * as React from "react";
import * as styles from "./styles.css";

import { classNames } from "../../tools/classnames";

export function TextInput(props: React.HTMLProps<HTMLInputElement>) {
    const { className, ...otherProps } = props;
    const classes = classNames(className, styles.input);

    return (
        <div className={styles.inputWrapper}>
            <input {...otherProps} className={classes} />
        </div>
    );
}
