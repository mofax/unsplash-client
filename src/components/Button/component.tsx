import * as React from "react";
import * as styles from "./styles.css";

import { Spinner } from "../Spinner/component";
import { classNames } from "../../tools/classnames";

type ButtonProps = {
    spinning?: boolean;
    type: "outline" | "pill";
};

export function Button(
    props: ButtonProps & React.HTMLProps<HTMLButtonElement>
) {
    const { spinning, children, ...otherProps } = props;
    return (
        <button
            {...otherProps}
            className={classNames(styles.button, styles[props.type])}
            type="button"
        >
            {children}
            {spinning ? <Spinner /> : null}
        </button>
    );
}

Button.defaultProps = {
    type: "pill",
};
