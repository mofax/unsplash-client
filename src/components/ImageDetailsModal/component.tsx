import * as React from "react";
import * as styles from "./styles.css";

import { GOOGLE_MAPS_KEY } from "../../../keys";
import { ImageStore } from "../../mobx/images";
import Marker from "../../svg/mapmarker";
import { Spinner } from "../Spinner/component";
import { classNames } from "../../tools/classnames";
import imageModalStore from "../../mobx/image-modal";
import { observer } from "mobx-react";

type ImageDetailsModalProps = {
    store?: any;
    imageStore: ImageStore;
};

export class ImageDetailsModalComponent extends React.PureComponent<
    ImageDetailsModalProps,
    {}
> {
    constructor(props: any) {
        super(props);
    }
    async componentDidMount() {
        if (!this.props.store.currentImage) {
            return;
        }

        const body = document.getElementById("body");
        body?.classList.add("noscroll");

        await this.props.imageStore.getSinglePhoto(
            this.props.store.currentImage.id
        );
    }
    componentWillUnmount() {
        const body = document.getElementById("body");
        body?.classList.remove("noscroll");
    }

    render() {
        const overlayClasses = [styles.overlay];

        if (this.props.store.open !== true)
            [overlayClasses.push(styles.hidden)];

        if (this.props.imageStore.fetchingSinglePhoto) {
            return (
                <div
                    className={classNames(overlayClasses)}
                    onClick={() => imageModalStore.toggle()}
                >
                    <div className={styles.detailsBox}>
                        <Spinner dark />
                    </div>
                </div>
            );
        }

        const image = this.props.imageStore.selectedPhoto;

        if (!image) {
            return null;
        }

        const imageUrl = image["urls"].regular;
        const user = image["user"];

        const userDetails = {
            name: user.name,
            username: user.username,
            thumbnail: user["profile_image"].small,
            link: user["links"].html,
        };

        let exifNode = null;

        if (image.exif) {
            const data = [
                "make",
                "model",
                "iso",
                "aperture",
                "exposure_time",
                "focal_length",
            ];

            exifNode = (
                <div>
                    {data.map((item: string) => {
                        if (!image.exif[item]) return null;

                        return (
                            <div>
                                <strong>{item}</strong>: {image.exif[item]}
                            </div>
                        );
                    })}
                </div>
            );
        }

        //embed google maps
        let mapsNode = null;
        const location = image.location;
        if (
            location &&
            location.position.latitude &&
            location.position.longitude
        ) {
            const center = `${location.position.latitude},${location.position.longitude}`;
            const link = `https://www.google.com/maps/embed/v1/place?key=${GOOGLE_MAPS_KEY}&center=${center}&q=${location.name}`;

            mapsNode = (
                <div className={styles.mapsWrapper}>
                    <iframe
                        className={styles.mapsIframe}
                        frameBorder="0"
                        style={{ border: 0 }}
                        src={link}
                        allowFullScreen
                    ></iframe>
                    <div className={styles.mapsLocationName}>
                        <Marker /> {location.name}
                    </div>
                </div>
            );
        }

        return (
            <div
                className={classNames(overlayClasses)}
                onClick={() => imageModalStore.toggle()}
            >
                <div className={styles.detailsBox}>
                    <img src={imageUrl} />
                    <div className={styles.userDetails}>
                        <img src={userDetails.thumbnail} />
                        <div>
                            <h2>{userDetails.name}</h2>
                            <a href={userDetails.link}>
                                @{userDetails.username}
                            </a>
                        </div>
                    </div>

                    <div className={styles.exif}>{exifNode}</div>
                    {mapsNode}
                </div>
            </div>
        );
    }
}

export const ImageDetailsModal = observer(ImageDetailsModalComponent);
