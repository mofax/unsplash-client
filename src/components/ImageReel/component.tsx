import * as React from "react";
import * as styles from "./styles.css";

import { Button } from "../Button/component";
import { Spinner } from "../Spinner/component";
import imageModalStore from "../../mobx/image-modal";
import imageStore from "../../mobx/images";

type ImageReelProps = {
    images: any[];
    loading?: boolean;
    onNextPage?: (direction: "next" | "previous") => void;
};

function ImageReel(props: ImageReelProps) {
    function imageClickHandler(imageItem: any) {
        imageStore.getSinglePhoto(imageItem.id);
        imageModalStore.toggle();
    }

    const imageMap = props.images.map((item) => {
        return (
            <div className={styles.imageWrapper} key={item["urls"].regular}>
                <img
                    onClick={imageClickHandler.bind(null, item)}
                    className={styles.image}
                    alt={item["alt_description"]}
                    src={item["urls"].regular}
                />
            </div>
        );
    });

    function navButtonsClickHandler(direction: "next" | "previous") {
        if (props.onNextPage) {
            props.onNextPage(direction);
        }
    }

    let node = null;
    if (props.loading) {
        node = <Spinner dark={true} width={100} height={100} />;
    } else if (!props.loading && props.images.length === 0) {
        node = (
            <div className={styles.error}>
                Sorry :( We don't have such photos!
            </div>
        );
    } else {
        node = (
            <>
                <div className={styles.reel}>{imageMap}</div>
                <div className={styles.footer}>
                    <Button
                        type="outline"
                        onClick={navButtonsClickHandler.bind(null, "previous")}
                    >
                        Previous
                    </Button>
                    <Button
                        type="outline"
                        onClick={navButtonsClickHandler.bind(null, "next")}
                    >
                        Next
                    </Button>
                </div>
            </>
        );
    }

    return <div className={styles.pageWrapper}>{node}</div>;
}

export { ImageReel };
