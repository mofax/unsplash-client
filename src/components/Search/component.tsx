import * as React from "react";
import * as styles from "./styles.css";

import { Button } from "../Button/component";
import { Logo } from "../../svg/logo";
import { SelectInput } from "../SelectInput/component";
import { TextInput } from "../TextInput/component";
import { classNames } from "../../tools/classnames";

type ImageSearchProps = {
    collections: any[];
    onSearch?: (searchTerm: string, collection: string) => void;
    searching?: boolean;
    collapsed: boolean;
};
export function ImageSearch(props: ImageSearchProps) {
    const [searchTerm, setSearchTerm] = React.useState("");
    const [selectedCollection, setCollection] = React.useState("");

    const [shake, setShake] = React.useState(false);

    function searchButtonClickHandler(e: React.MouseEvent<HTMLButtonElement>) {
        e.preventDefault();
        if (!searchTerm) {
            setShake(true);
            setTimeout(() => {
                setShake(false);
            }, 1000);
        } else {
            if (props.onSearch && props.onSearch.call) {
                props.onSearch(searchTerm, selectedCollection);
            }
        }
    }

    const collectionOptions = props.collections.map((c) => {
        return { label: c.title, value: c.id };
    });

    const wrapperClassnames = [styles.wrapper];
    if (props.collapsed) {
        wrapperClassnames.push(styles.wrapperCollapsed);
    }
    return (
        <div className={classNames(wrapperClassnames)}>
            <div className={styles.searchbox}>
                <div className={styles.logo}>
                    <Logo />
                    <br></br>
                    <div className={styles.text}>
                        <strong>image</strong>&nbsp;search
                    </div>
                </div>
                <div className={styles.textinput}>
                    <TextInput
                        className={shake && styles.shake}
                        value={searchTerm}
                        onChange={(e: React.FormEvent<HTMLInputElement>) =>
                            setSearchTerm((e.target as HTMLInputElement).value)
                        }
                        placeholder="Query"
                    />
                    <SelectInput
                        placeholder="Collections"
                        options={collectionOptions}
                        value={selectedCollection}
                        onChange={(arg) => {
                            setCollection(arg.value);
                        }}
                    ></SelectInput>
                </div>
                <Button
                    onClick={searchButtonClickHandler}
                    disabled={props.searching}
                    spinning={props.searching}
                >
                    Search
                </Button>
            </div>
        </div>
    );
}
