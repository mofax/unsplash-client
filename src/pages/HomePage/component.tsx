import * as React from "react";

import { ImageDetailsModal } from "../../components/ImageDetailsModal/component";
import { ImageReel } from "../../components/ImageReel/component";
import { ImageSearch } from "../../components/Search/component";
import imageDetailsStore from "../../mobx/image-modal";
import imageStore from "../../mobx/images";
import { observer } from "mobx-react-lite";

function HomePageComponent() {
    async function searchImages(term: string, collection?: string) {
        await imageStore.search(term, collection);
    }

    async function navigatePage(direction: "next" | "previous") {
        if (direction === "next") {
            imageStore.gotoNextPage();
        } else {
            imageStore.gotoPreviousPage();
        }
    }

    React.useEffect(() => {
        imageStore.fetchCollections();
    }, []);

    let reelNode = null;
    if (!!imageStore.searchResults) {
        reelNode = (
            <>
                <ImageReel
                    loading={imageStore.searching}
                    images={imageStore.searchResults?.results}
                    onNextPage={navigatePage}
                />
                <ImageDetailsModal
                    store={imageDetailsStore}
                    imageStore={imageStore}
                />
            </>
        );
    }
    return (
        <div>
            <ImageSearch
                collapsed={!!imageStore.searchResults}
                collections={imageStore.collections}
                onSearch={searchImages}
                searching={imageStore.searching}
            />
            {reelNode}
        </div>
    );
}

export const HomePage = observer(HomePageComponent);
