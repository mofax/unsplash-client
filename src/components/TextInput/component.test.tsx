import * as React from "react";

import { TextInput } from "./component";
import renderer from "react-test-renderer";

describe("<TextInput>", function () {
    it("renders correctly", () => {
        const tree = renderer.create(<TextInput />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
