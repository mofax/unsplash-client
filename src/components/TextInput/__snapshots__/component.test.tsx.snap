// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput> renders correctly 1`] = `
<div>
  <input
    className=""
  />
</div>
`;
